proposals/2023/04-21_build-init-log.md

```diff
@@ -0,0 +1,292 @@
+
+```
+- CreateBuildInitLog:  POST   /api/v1/repos/:org/:repo/builds/:build/init/logs
+- GetBuildInitLog:     GET    /api/v1/repos/:org/:repo/builds/:build/init/logs
+- UpdateBuildInitLog:  PUT    /api/v1/repos/:org/:repo/builds/:build/init/logs
+- DeleteBuildInitLog:  DELETE /api/v1/repos/:org/:repo/builds/:build/init/logs
+```
+
+The API path ends up being the only place that needs this "init" string. This avoids ambiguity between retrieve all logs (including step and service logs) and the one `Log` associated only with the `Build` (not a `Step` or `Service`).
+
+These endpoints need to be registered in the router<!-- `LogBuildInitHandlers` in `router/log.go` -->, but no new middleware is required.
+
+Also include the endpoints in the mock server.
+
+#### Server - Compiler
+
+The compiler needs to deprecate and stop adding the `InitStage` and `InitStep`. That includes:
+- remove `compiler/native/initialize{,_test}.go`,
+- remove `Init*` from the `Engine` interface in `compiler/engine.go`,
+- remove the `Init*()` calls from `compiler/native/compile.go`, and
+- remove the magic `"init"` string special-casing in `compiler/native/validate.go`.
+
+Backwards compatibility is a concern with the compiler. Referencing old builds/pipelines should work just fine as it includes the "init" step and stage, thus preserving references to the underlying log. But, recompiling the pipeline will create a slightly different pipeline--one without the injected "init" stage and/or step. The worker will no longer handle the old pipeline with the injected "init" step, so any re-runs MUST re-compile the pipeline.
+
+Does re-running a build ALWAYS re-compile the pipeline?
+
+#### Server - Queue
+
+Once the worker is upgraded to stop special-casing the `"init"` and `"#init"` strings, any queued builds that have that `"init"` pseudo-`Step` will fail.
+
+Also, when upgrading, we need to ensure that:
+- the queue is empty, or
+- all queued builds get re-compiled (or more particularly `item.Pipeline` which is a `types.pipeline.Build` and includes the injected init step) before execution in the worker starts.
+
+To make the worker backwards compatible, could we trigger the rebuild in `server.queue.redis.Pop()` after the item is created?
+Or perhaps we'll just need to include queued build migrations in the migration script.
+
+### Worker
+
+Save init logs to the Build `Log` instead of the magic init step.
+Nothing in the worker should check for these magic strings any more: `"init"`, `"#init"`
+
+We might need something that rejects an old build when popped off the queue if it is an older version that includes the injected "init" `Step`.
+
+We might need to spread this change over a couple of versions.
+- In one version, we start logging to the Build `Log` and deprecate support for the magic "init" step.
+- In the next (or a future) version, we remove support for the magic string checking.
```

Comment:
This sounds good to me 👍 

FWIW - I'm not opposed to doing this all in a single version assuming we document it as a breaking change.